### PR TITLE
Allow DownloadWorkManager to have Bundle based requests along with url based ones.

### LINKED ISSUE #1888: Allow applications to download resources using Bundle requests.
**Is your feature request related to a problem? Please describe.**
During the sync process, the application may want download users in a Bundle request, but the current engine api only allows url based download requests. 
For a Bundle request example, see https://www.hl7.org/fhir/bundle-request-medsallergies.json.html .

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Would you like to work on the issue?**
Please state if this issue should be assigned to you or who you think could help to solve this issue.

### PR DESCRIPTION
**Pull request by aditya-07**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1888

**Description**
1. Added `Request` type to `DownloadWorkManager` to support Url and Bundle based download requests.
2. Added an overloaded `download(Bundle)` api in `DataSource`.
3. Updated tests for `DownloaderImpl`.

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one:  Feature

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: engine/src/test/java/com/google/android/fhir/sync/download/DownloaderImplTest.kt [hunk 1]
```diff
@@ -36,302 +44,168 @@ import org.robolectric.RobolectricTestRunner
...
-              else -> OperationOutcome()
-            }
+        }
+      }
+
+    private val testDataSource: DataSource =
+      object : DataSource {
+        override suspend fun download(path: String): Resource {
+          return when {
+            path == "Patient" ->
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/sync/download/DownloaderImplTest.kt:128]: i actually think these would make sense to be part of the test cases - shouldn't be hardcoded here.

File: engine/src/main/java/com/google/android/fhir/sync/DataSource.kt [hunk 2]
```diff
@@ -31,6 +32,13 @@ internal interface DataSource {
    */
   suspend fun download(path: String): Resource
 
+  /**
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/sync/DataSource.kt:35]: another option is to have get and post functions to be implemented by data source impls, but also expose download and upload using bundle (both calling the underlying post function)

**Review by jingtang10 — APPROVED** [review]

File: engine/src/main/java/com/google/android/fhir/sync/DownloadWorkManager.kt [hunk 3]
```diff
@@ -31,11 +32,11 @@ interface DownloadWorkManager {
    * Returns the URL for the next download request, or `null` if there is no more download request
    * to be issued.
    */
-  suspend fun getNextRequestUrl(context: SyncDownloadContext): String?
+  suspend fun getNextRequest(context: SyncDownloadContext): Request?
 
   /* TODO: Generalize the DownloadWorkManager API to not sequentially download resource by type (https://github.com/google/android-fhir/issues/1884) */
```
Comment (omarismail94) [inline engine/src/main/java/com/google/android/fhir/sync/DownloadWorkManager.kt:37]: is this TODO still relevant?

File: engine/src/test/java/com/google/android/fhir/sync/download/DownloaderImplTest.kt (hunk 1 above)
Comment (aditya-07) [reply engine/src/test/java/com/google/android/fhir/sync/download/DownloaderImplTest.kt]: Done

File: engine/src/test-common/java/com/google/android/fhir/resource/TestingUtils.kt [hunk 4]
```diff
@@ -193,8 +199,12 @@ class TestingUtils constructor(private val iParser: IParser) {
       throw Exception(hugeStackTraceMessage)
     }
 
+    override suspend fun download(bundle: Bundle): Resource {
+      throw Exception("Posting Download Bundle failed...")
```
[the remark this answers is not in the document]
Comment (aditya-07) [reply engine/src/test-common/java/com/google/android/fhir/resource/TestingUtils.kt]: `ResourceSyncException` is unfortunately data class. Maybe throwing `java.net.SocketTimeoutException` is a better real life scenario.

File: engine/src/main/java/com/google/android/fhir/sync/remote/RemoteFhirService.kt [hunk 5]
```diff
@@ -36,6 +36,8 @@ internal interface RemoteFhirService : DataSource {
 
   @GET override suspend fun download(@Url path: String): Resource
```
[the remark this answers is not in the document]
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/sync/remote/RemoteFhirService.kt]: Done. `RemoteFhirService` now has a http interface like apis i.e. `get` and `post`.

File: engine/src/main/java/com/google/android/fhir/sync/DataSource.kt (hunk 2 above)
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/sync/DataSource.kt:35]: Done.

File: engine/src/main/java/com/google/android/fhir/sync/DownloadWorkManager.kt (hunk 3 above)
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/sync/DownloadWorkManager.kt:37]: Its still relevant as the return type of `getSummaryRequestUrls` still has the assumption that the download is per resource type.

File: engine/src/main/java/com/google/android/fhir/sync/remote/RetrofitBasedRemoteDataSource.kt [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.android.fhir.sync.remote
+
+import com.google.android.fhir.sync.DataSource
+import org.hl7.fhir.r4.model.Bundle
+
+internal class RetrofitBasedRemoteDataSource(private val remoteFhirService: RemoteFhirService) :
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/sync/remote/RetrofitBasedRemoteDataSource.kt:22]: doc

File: engine/src/main/java/com/google/android/fhir/sync/remote/RemoteFhirService.kt [hunk 7]
```diff
@@ -32,11 +31,11 @@ import retrofit2.http.POST
 import retrofit2.http.Url
 
 /** Interface to make http requests to the FHIR server. */
-internal interface RemoteFhirService : DataSource {
+internal interface RemoteFhirService {
```
[the remark this answers is not in the document]
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/sync/remote/RemoteFhirService.kt]: Done

File: engine/src/main/java/com/google/android/fhir/sync/remote/RetrofitBasedRemoteDataSource.kt (hunk 6 above)
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/sync/remote/RetrofitBasedRemoteDataSource.kt]: Removed as per our discussion.

**Pull request merged**